Add support for --cloud-init + --user-mode-networking

cfergeau opened the pull request:
This PR will allow to use --cloud-init and --user-mode-networking simultaneously on hyper-v.
It supersedes https://github.com/cfergeau/podman/pull/36. We could either go with this PR, or add the first 2 commits to #36, and then take this one.
Note: I haven’t tested this yet.

lstocchi committed:
cloud-init: generate user-data config file based on settings on hyperv

if users provide their custom user-data we use theirc file. However,
user-mode networking requires some customization on the VM and we cannot
expect the user to provide such values. So, if users enable user-mode
networking and they provide a custom cloud-init, we update it to apply
the settings needed for user-mode networking to work fine.

if users provide a user-data file and they rely on system networking, we
keep their user-data as is.

If users do not provide any cloud init configuration, it creates a
default one.

Signed-off-by: lstocchi <lstocchi@redhat.com>

cfergeau committed:
cloud-init: Fix linting issues

`range` over a `nil` array is ok in go, so the `nil` check is not
needed.
The `User` type was also redundant and flagged by golangci-lint.

Signed-off-by: Christophe Fergeau <cfergeau@redhat.com>

cfergeau committed:
cloud-init: Add UserData:MarshalMultiPart helper

This will allow better separation between the high level decisions
(usermode networking enabled/disabled and --cloud-init used or not) and
the low-level byte details

Signed-off-by: Christophe Fergeau <cfergeau@redhat.com>

cfergeau committed:
cloud-init: Move `UserData` related code to userdata.go

The end goal is to make the rest of the code easier to follow by
grouping most of the low-level code generating cloud-init data files to
this userdata.go file.

Signed-off-by: Christophe Fergeau <cfergeau@redhat.com>

cfergeau committed:
cloud-init: windows: Simplify generateUserData

This adds a defaultHypervUserData helper to only keep high level logic
in generateUserData and make the different cases easier (with/without
user-mode networking, with/without custom user-data) to follow.
There should be no behaviour changes.

Signed-off-by: Christophe Fergeau <cfergeau@redhat.com>

cfergeau committed:
cloud-init: Refactor existing code

This is in preparation for the next commit which will add support for
using --cloud-init and --user-mode-networking together.
This commit should introduce no functional changes, just code movement.

This is based on work from lstocchi <lstocchi@redhat.com>

Signed-off-by: Christophe Fergeau <cfergeau@redhat.com>

lstocchi commented on pkg/machine/cloudinit/cloudinit_windows.go:
```diff
@@ -8,73 +8,93 @@ import (
-	yamlBytes = append([]byte(headerLine), yamlBytes...)
-	return yamlBytes, nil
+	// if user has provided a custom user-data but we're not on Hyper-V/user-mode networking, return it as-is
+	if userModeNetworking {
```
```suggestion
	if !userModeNetworking {
```

cfergeau replied:
Thanks for the catch, this is addressed in the latest push.

lstocchi committed:
remove newline from beginning HyperVVsockNMConnection

it removes the initial newline char from the string
HyperVVsockNMConnection.
The newline creates no issues for ignition but break the configuration
when using cloud-init. This fixes the problem.

lstocchi committed:
define cloud-init merging strategy

when merging two cloud-init config files like the case where one user-data is passed by the user and one is
created by us to support user-mode networking, we need to make sure we
define a merging strategy, otherwise cloud-init will just use the last
clause by overwriting the old one.

For example if there are two files like

runcmd:
  - bash1
  - bash2

runcmd:
  - bash3
  - bash4

if we do not define a merging strategy, cloud-init will only run bash3
and bash4.

This could be an issue if user passes a user-data file containing same
clauses we use in our user-mode networking config file (like runcmd).

This path defines a strategy to merge lists of value and do not do
anything with string

Signed-off-by: lstocchi <lstocchi@redhat.com>

lstocchi committed:
use actual path where gvforwarder is copied by cloud-init

Signed-off-by: lstocchi <lstocchi@redhat.com>

lstocchi reviewed (commented):
Found some errors (mostly related to my work after last rebase + one new thing). I added 3 new commits in old PR so it's easier for you to view the changes https://github.com/cfergeau/podman/pull/36. 
Give them a look and add them here if you're ok with the changes. Thanks :)

lstocchi commented on pkg/machine/cloudinit/cloudinit_windows.go:
```diff
@@ -8,73 +8,93 @@ import (
-	if mc.HyperVHypervisor != nil && mc.HyperVHypervisor.UserModeNetworking {
-		netUnitFile, err := hutil.CreateNetworkUnit(mc.HyperVHypervisor.NetworkVSock.Port)
+	userData.RunCmd = []string{
+		"install -o root -g root -m 0755 /mnt/gvforwarder /usr/local/bin/gvforwarder",
```
the code works fine now but i found an issue with a rhel bootc image i created some days ago. 
We cannot write on /usr ... so this won't work. Should we move the path somewhere else? like in /var? But then, on RHEL, if we move the file we must tell selinux it's fine to be run by setting the context

cfergeau replied:
I think this can be addressed in a follow-up PR. user-mode networking + hyperv + bootc won’t work for now, we can say it’s a known issue.

lstocchi replied:
Ok for me

cfergeau replied:
https://github.com/crc-org/macadam/issues/305

lstocchi reviewed (commented):
Works fine now, thanks. There is just a problem with bootc images. Not sure if you want to have a look here or another PR. Fine to merge this as it is for me
